### ISSUE TITLE
Unreliable video playback in web story

### ISSUE THREAD

**Issue by alankent** [ampproject/amphtml#37190]
### Description

I came back to a web story I have not viewed for a few weeks (https://extra-ordinary-web-stories.web.app/ep1/). Each page is a video clip, nothing more. Many of the pages came up as white (no video shown). On the console I see messages such as

```Failed to load resource: net::ERR_CACHE_OPERATION_NOT_SUPPORTED```

![image](https://user-images.githubusercontent.com/5702163/145705394-7a846731-d999-478d-b6a9-ebe57e0427bc.png)

If I right click on URL to the mp4 file and open in a new tab, it plays fine (no problems). After a while of going forwards and back and doing page reloads the videos start working. I then go forward a few pages fine, then the errors come back again. Not sure why, hard to repeat reliably.

### Reproduction Steps

https://extra-ordinary-web-stories.web.app/ep1/

But it seems to be a caching issue - it does not reliably fail for me. An incognito window seemed to be more reliable than a normal browser tab, so maybe its some service worker cache or similar half timed out? (Video gone, but it thinks it still in cache??)

### Relevant Logs

```shell
Failed to load resource: net::ERR_CACHE_OPERATION_NOT_SUPPORTED
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
/ep1/ep1-10-050.mp4:1 Failed to load resource: net::ERR_CACHE_OPERATION_NOT_SUPPORTED
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
/ep1/ep1-10-070.mp4:1 Failed to load resource: net::ERR_CACHE_OPERATION_NOT_SUPPORTED
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
/ep1/ep1-10-090.mp4:1 Failed to load resource: net::ERR_CACHE_OPERATION_NOT_SUPPORTED
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
log.js:295 [video-manager] This video is too short for `video-percentage-played`. Reports may be innacurate. For best results, use videos over 5 seconds long. <amp-video width=​"868" height=​"1462" layout=​"fill" autoplay noaudio class=​"i-amphtml-element i-amphtml-layout-fill i-amphtml-layout-size-defined i-amphtml-disable-mediasession i-amphtml-poolbound i-amphtml-media-component i-amphtml-video-interface i-amphtml-built i-amphtml-layout" i-amphtml-layout=​"fill" preload=​"auto">​…​</amp-video>​
Gb @ log.js:295
```

### Browser(s) Affected

Chrome

### OS(s) Affected

Windows (I have not tested others - chrome on iPad seems okay)

### Device(s) Affected

Desktop PC with Windows 10 and Chrome is where I tested mainly

**Comment by alankent** [comment]
Oh, just got iPad to fail as well with "Video failed to play" in bottom right corner. (Note the top nav markers not being at the top of the screen on an iPad is a separate previously reported bug.)

![image0](https://user-images.githubusercontent.com/5702163/145705703-c7a8b7a1-1e74-4b52-9706-29cb61328bbb.png)

Pressing "Reload" made the video start loading correctly, and all following pages worked. 

All the files (including the video files are mp4 files) are hosted in Firebase.

**Comment by alankent** [comment]
Oh, and viewing the same story again 10 mins later on iPad, other pages that previously worked failed to load (white pages) and using "reload" did not make them appear either. Clicking forward and back a few times and it started working again. One page I clicked reload 4 times and then it started working.

**Comment by gmajoulet** [comment]
Happy to help debugging this, but I'd like to note that it sounds likely not Stories format code related. We haven't seen any increased error rate in our dashboards.
I haven't been able to repro on any of my devices but will keep trying.

**Comment by alankent** [comment]
Trying again now. Worked perfectly on laptop this time. Failed once on iPad (but I don't have hooked up to Mac to view console logs). It was failing many times yesterday on multiple devices. (Scratches head.)

I did see some different "cancelled" messages on Mac Chrome console. Example for  https://extra-ordinary-web-stories.web.app/ep1/ep1-30-240.mp4

- First request is GET for MP4 file is `range: bytes=0-`
- Second request is GET is `bytes 0-425414/425415`
- Next is GET `bytes 393216-425414/425415`
- Next is GET ` bytes=98304-` which is cancelled
- Final is GET `bytes=98304-` which succeeds

Note that this may be normal and intended behavior.
![image](https://user-images.githubusercontent.com/5702163/145896041-ab293b8c-207c-45d9-9183-004dce53583b.png)

Hmmm. Flipping backwards and forwards on ipad it starts failing more over time. (Pages that used to work are stopping after forward/back multiple times.) Feels like a potential race condition in the service worker (or something funky like that). I will try to get access to ipad browser console.

**Comment by gmajoulet** [comment]
We haven't experimented much with service workers for Stories ourselves, please let us know if you find what the issue is.

**Comment by alankent** [comment]
Got console on iPad going. I saw a "decode error" on some files. However when I open the files directly in a tab they seem fine? No errors from console? (I am stumbling around Safari console / debugging tools.)

![image](https://user-images.githubusercontent.com/5702163/145898651-5de3b8b1-caba-4812-9b21-f1a1948fc352.png)

Okay, this I don't understand.
![image](https://user-images.githubusercontent.com/5702163/145900035-92fb8458-b7a9-4ca1-8785-c3efcce02a5a.png)
I hit the back button a few times (page 190 back to 110) - no network requests to fetch images, white screen ("Video unable to play"). I then hit forward button. No requests for those videos its failing on (from Safari console on iPad connected to Mac via cable. (I was expecting to see 120, 130, 140 etc. - I skip some numbers, but nothing was appearing.)

It is almost like the browser after seeing a previous error for a URL is no longer fetching it again? I was expecting to see a network request for the "white" pages.

I might see if I can turn off service workers in Safari and try again to see if I can narrow it down that way.

**Comment by alankent** [comment]
I don't have my own service worker - I was just wondering if it could be a caching timing issue (I thought web stories registered a service worker to cache content behind the scenes). If that is not the case, I am just wasting my time trying to turn them off! ;-)

No idea what is happening.


Hmmm. Tried again just now on Windows instead of Mac, still getting the same error. Going to reboot Windows next. (Just restarted Chrome, updated to latest version, etc).

**Comment by alankent** [comment]
Ah! There is a chrome bug that might be the problem. Multiple concurrent requests too the same URL specifying byte ranges fails at times. That does not describe the iPad problem, but does explain the Windows Chrome problem which is still occurring. 
https://bugs.chromium.org/p/chromium/issues/detail?id=770694

**Comment by gmajoulet** [comment]
To clarify, the bug happens when you specific the same URL for multiple videos within a Story?

**Comment by alankent** [comment]
No. Each page is a unique video clip. The video clip is the only thing on each page.

```
<amp-story-page id="ep1-10-032">
  <amp-story-grid-layer preset="2021-background" template="fill">
    <amp-video src="ep1-10-032.mp4" width="868" height="1462" layout="fill" autoplay noaudio></amp-video>
  </amp-story-grid-layer>
</amp-story-page>
<amp-story-page id="ep1-10-034">
  <amp-story-grid-layer preset="2021-background" template="fill">
    <amp-video src="ep1-10-034.mp4" width="868" height="1462" layout="fill" autoplay noaudio></amp-video>
  </amp-story-grid-layer>
</amp-story-page>
<amp-story-page id="ep1-10-036">
  <amp-story-grid-layer preset="2021-background" template="fill">
    <amp-video src="ep1-10-036.mp4" width="868" height="1462" layout="fill" autoplay noaudio></amp-video>
  </amp-story-grid-layer>
</amp-story-page>
```

The Chrome bug generating the same error message is when the web browser makes multiple parallel requests to the same video file. Looking at the network logs you can see the browser do lots of requests for the same video file (including some cancelations of requests) to play the video.

If you wanted a guess, I think web stories start playing the video on a page and then tells the player "start again using this URL" - there is a delay between the video player component being active and the player being given the URL for the current page. I have reported in a previous bug (https://github.com/ampproject/amphtml/issues/35480) that you can often see the first few frames of the previous page video play on the new page, then it starts playing the correct video. I am wondering if there is some strange race condition where the video player component is being told the URL to play multiple times, and the player component is restarting (not optimized to know its the same URL) so it starts to cancel the current video and tell it to start a new video clip (but its the same video clip actually) - resulting in multiple concurrent requests. (Or multiple player objects are being created and destroyed - I have not looked at the code.) I don't see why the web browser is not just opening the URL and streaming it to the player in a single request rather than fetching lots of byte ranges. My suspicion is JavaScript calls to the video player are making it stutter and restart, which is also triggering the Windows Chrome bug now.

**Comment by alankent** [comment]
As another experiment, I created a single <video> element on a page to see what HTTP requests it made to fetch the video file. As I would have expected, one HTTP request per mp4 file. 

```
<html>
  <body>
    <video width=320 height=240 controls>
      <source src="https://extra-ordinary-web-stories.web.app/ep1/ep1-10-037.mp4" type="video/mp4">
    </video>
  </body>
</html>
```

The web story however is making multiple requests, with cancel requests. So I created a simple React app that changes the src attribute to different videos. It is immediately doing multiple HTTP requests (typically 3 per video when I click the 'next video' button). So it is something about the runtime DOM manipulation causing multiple HTTP requests. Not sure there is a workaround possible or not.
